mephisto/utils/metrics.py

```diff
@@ -0,0 +1,206 @@
+METRICS_DIR = os.path.join(get_root_dir(), "metrics")
+PROMETHEUS_DIR = os.path.join(METRICS_DIR, "prometheus")
+PROMETHEUS_EXECUTABLE = os.path.join(PROMETHEUS_DIR, "prometheus")
+PROMETHEUS_CONFIG = os.path.join(PROMETHEUS_DIR, "prometheus.yml")
+GRAFANA_DIR = os.path.join(METRICS_DIR, "grafana")
+GRAFANA_EXECUTABLE = os.path.join(GRAFANA_DIR, "bin", "grafana-server")
+
+
+def _server_process_running(pid):
+    """Check on the existing process id"""
+    try:
+        os.kill(pid, 0)
+    except OSError as err:
+        if err.errno == errno.ESRCH:
+            # ESRCH == No such process
+            return False
+        elif err.errno == errno.EPERM:
+            # EPERM clearly means there's a process to deny access to
+            return True
+        else:
+            raise err
+    else:
+        return True
+
+
+def _get_pid_from_file(fn):
+    """Get the PID from the given file"""
+    with open(fn) as pid_file:
+        pid = int(pid_file.read().strip())
+    return pid
+
+
+def start_metrics_server(args: Optional["DictConfig"] = None):
+    """
+    Launch a metrics server for the current job. At the moment, defaults to launching on port 3031.
+
+    Unfortunately this means we are only able to check metrics on one job at a time for now.
+    Future work will extend our metrics logging configuration.
+    """
+    try:
+        start_http_server(3031)
+    except Exception as e:
+        logger.exception(
+            "Could not launch prometheus metrics client, perhaps a process is already running on 3031? "
+            "Mephisto metrics currently only supports one Operator class at a time at the moment",
+            exc_info=True,
+        )
+
+
+def launch_prometheus_server(args: Optional["DictConfig"] = None) -> bool:
+    """
+    Launch a prometheus server if one is not already running (based on having an expected PID)
+
+    Returns success or failure
+    """
+    if os.path.exists(PROMETHEUS_PID_FILE):
+        r = requests.get("http://localhost:9090/")
+        if not r.ok:
+            logger.warning(
+                "Prometheus PID existed, but server doesn't appear to be up."
+            )
+            if _server_process_running(_get_pid_from_file(PROMETHEUS_PID_FILE)):
+                logger.warning(
+                    "Prometheus server appears to be running though! exiting as unsure what to do..."
+                )
+                raise Exception("Prometheus server running but inaccessible")
+            else:
+                logger.warning(
+                    "Clearing prometheus pid as the server isn't running. "
+                    "Use `shutdown_prometheus_server` in the future for proper cleanup."
+                )
+        else:
+            logger.info("Prometheus server appears to be running at 9090")
+            return True
+    if not os.path.exists(PROMETHEUS_EXECUTABLE):
+        logger.warning(
+            f"Cannot collect and display metrics without installing metrics. "
```

Comment:
Should this read "without installing Prometheus" instead of "without installing metrics"?

Yeah I think I would rephrase this as more of a positive offering than a lack of a feature. For example a message like:

"Mephisto supports rich run-time metrics visualization through Prometheus. If you'd like to use this feature, use the scripts in the metrics directory <...> to install"

Addendum: What happens when people install Mephisto via pip, they won't have access to the metrics directory in that case then right? In that case mentioning the metrics dir could be confusing to them.

Reply:
> What happens when people install Mephisto via pip, they won't have access to the metrics directory in that case then right

They do have access, but it may not be where they expect. This is why I get `METRICS_DIR`. I imagine we'll want to add things to the `mephisto` cli to access metrics in the future though.

Reply:
Wondering if we could quickly augment the CLI to support `mephisto script view_metrics` which will use the logic here: https://stackoverflow.com/a/5790381 to run the script file from within the package.

Reply:
Added scripts to the CLI:
```
>>> mephisto metrics
Usage: mephisto metrics <install|launch|cleanup>
install: Installs prometheus and grafana to /Users/jju/mephisto/metrics
view: Launches a Prometheus and Grafana server, and shuts down on exit
cleanup: Shuts down prometheus and grafana resources that may have persisted
```